### ISSUE TITLE
ROPE is necessary in VQ Stoks, but position is not provided

### ISSUE THREAD

**Issue by harish-kamath** [collabora/WhisperSpeech#64]
https://github.com/collabora/WhisperSpeech/blob/b6fc87c9e1ffd28b0fef71bc25896ce323a76fa8/whisperspeech/vq_stoks.py#L340

Getting an error here - but upon digging in I don’t know what the solution would be, and don’t know why this doesn’t happen for every inference. 

Essentially, if ROPE is true (and it is by default for VQ stoks), then you expect positional information to be passed here otherwise you get an error (None can’t multiply float):

https://github.com/collabora/WhisperSpeech/blob/b6fc87c9e1ffd28b0fef71bc25896ce323a76fa8/whisperspeech/modules.py#L108

But it isn’t as you can see from the initial call to ‘out_blocks’. I see that the ‘upgrade’ method in the tunable for VQ stoks turns off rope - but don’t see this method called anywhere in the repo. Should rope be off for VQ stoks?

**Comment by jpc** [comment]
Hey, that's a good catch, sorry about that. This seems to be a regression related to the optimization work we did recently. Since `vq_stoks` is not normally used during inference I did not notice that I broke the API.

Some context: `x_positions` is only relevant when doing inference with a `kv_cache`. Otherwise we are just passing `arange(0,seq_len)` around. Of the top of my head you should be able to get it to work by adding `torch.arange(0,x.shape[-1], device=x.device)` as a second argument to `out_blocks`.

It would be also worth checking if these additional `position` arguments interfere with training (they might make it slower?).

`upgrade` is for backwards compatibility of old models – the Tunables default values are updated as we discover the optimal settings. Without `upgrade` old models would inherit these defaults for all new settings that were added after they were trained and stop working. `upgrade` is used on deserialized JSON to detect missing keys and fill them with backwards compatible values.

**Comment by harish-kamath** [comment]
Oh woah I didn’t realize the model wasn’t used in inference as well. So does the S2A model also only take the token IDs? In other words during inference we never actually use whisper embeddings?

**Comment by jpc** [comment]
Both models use the semantic token ids but we also load the frozen token embeddings from the vq_stoks model and use linear projection layers to project them into the model dimension (and "unproject" for logits in the T2S model).

This helped make sure the T2S and S2A actually share a common semantic token embedding space and don't overfit. I borrowed this trick from Vocos which does the same thing with Encodec token embeddings.
